### PR TITLE
o/snapstate: use refresh hints to continue auto-refresh after monitoring

### PR DESCRIPTION
**Pull request by MiguelPires**
This addresses the [issue of Firefox auto-refreshes not being continued](https://bugs.launchpad.net/snapd/+bug/2019158) after the user stops the snap. The root cause is that the second auto-refresh doesn't always get a consistent response from the store. That is, the first auto-refresh might get a new revision but, if it's inhibited by a running snap, the next auto-refresh might not. This only happens for Firefox (to my knowledge) because the store only throttles Firefox, to reduce the load on its servers. This PR uses the refresh hints written during the auto-refresh (added [here](https://github.com/snapcore/snapd/pull/12946)) to continue the auto-refresh without relying on the store to return the same information.

### CODE REVIEW

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -834,16 +834,10 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	}
 
 	targetFn := snapsup.MountFile()
-	dlOpts := &store.DownloadOptions{
-		// pre-downloads are only triggered in auto-refreshes
-		Scheduled: true,
-		RateLimit: autoRefreshRateLimited(st),
-	}
-
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:842]: also not sure why this change?

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -892,113 +886,196 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
...
 		return fmt.Errorf("cannot monitor for snap closure: %w", err)
 	}
 
 	// notify the user about the blocked refresh
 	asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
 
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		return fmt.Errorf("cannot get refresh-candidates: %v", err)
+	}
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:918]: refreshHints doesn't seem to be used in this function itself?

File: overlord/snapstate/handlers.go [hunk 3]
```diff
@@ -892,113 +886,196 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
...
+
 	abort := make(chan bool, 1)
-	monitoredSnaps[refreshInfo.InstanceName] = abort
-	updateMonitoringState(st, monitoredSnaps)
+	if err := addMonitoring(st, snapName, abort); err != nil {
+		return fmt.Errorf("cannot save monitoring state: %v", err)
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:922]: should this be an error or only something that we log?

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -1707,7 +1707,7 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, re
 
 	names = strutil.Deduplicate(names)
 
-	refreshOpts := &store.RefreshOptions{Scheduled: flags.IsAutoRefresh}
+	refreshOpts := &store.RefreshOptions{Scheduled: flags.IsAutoRefresh && !flags.IsContinuedAutoRefresh}
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply overlord/snapstate/snapstate.go:1710]: True, I'll remove this

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:922]: I think it makes sense to bubble up the error and fail the pre-download change since we're not sure whether we saved an abort channel or monitoring info. If we keep going, then we might not have state to recover from restarts, abort the monitoring or determine whether there's a monitoring goroutine already. So that might result in weird scenarios and bugs

File: overlord/snapstate/handlers.go (hunk 2 above)
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:918]: Ah it's a leftover from when it got moved into the function below. I'll remove it

File: overlord/snapstate/handlers.go (hunk 1 above)
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:842]: I think I removed it when the field was AutoRefresh, I'm not sure the scheduled header makes sense for the pre-download. In particular, since we now know that in other cases it's used to omit a change (which I assume would never apply to a download but still). Do you think I should leave it?

Comment (pedronis) [reply overlord/snapstate/handlers.go:842]: it is still part of an auto-refresh, also we for sure want the ratelimit

File: overlord/snapstate/handlers.go (hunk 3 above)
Comment (pedronis) [reply overlord/snapstate/handlers.go:922]: I meant log and return, not log and continue. But maybe that was clear?

Comment (MiguelPires) [reply overlord/snapstate/handlers.go:922]: Ah sure, yeah I can change that

File: overlord/snapstate/handlers.go [hunk 5]
```diff
@@ -736,7 +736,7 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	st.Lock()
 	perfTimings := state.TimingsForTask(t)
 	snapsup, theStore, user, err := downloadSnapParams(st, t)
-	if snapsup != nil && snapsup.IsAutoRefresh {
+	if snapsup != nil && snapsup.IsAutoRefresh && !snapsup.IsContinuedAutoRefresh {
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:739]: Removed

File: overlord/snapstate/handlers.go [hunk 6]
```diff
@@ -753,7 +753,7 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	targetFn := snapsup.MountFile()
 
 	dlOpts := &store.DownloadOptions{
-		Scheduled: snapsup.IsAutoRefresh,
+		Scheduled: snapsup.IsAutoRefresh && !snapsup.IsContinuedAutoRefresh,
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:756]: Removed

**Review by pedronis — COMMENTED** [review]
thanks for the changes, question about what should happen when a snap disappears from refresh-candidates. Also should we check if changing refresh-candidates drop snaps, and should we stop their monitoring there?

File: overlord/snapstate/autorefresh.go [hunk 7]
```diff
@@ -98,16 +101,23 @@ func (rc *refreshCandidate) InstanceName() string {
...
 	if err := Get(st, rc.InstanceName(), &snapst); err != nil {
 		return nil, nil, err
 	}
-	return &rc.SnapSetup, &snapst, nil
+
+	snapsup := &rc.SnapSetup
+	if globalFlags != nil {
+		snapsup.Flags.IsAutoRefresh = globalFlags.IsAutoRefresh
+		snapsup.Flags.IsContinuedAutoRefresh = globalFlags.IsContinuedAutoRefresh
+	}
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply overlord/snapstate/autorefresh.go:118]: Added a test for this

**Comment by MiguelPires** [comment]
> Also should we check if changing refresh-candidates drop snaps, and should we stop their monitoring there?
 
I added some code to the `refreshhints.go` that aborts the monitoring routines of snaps for which hints were dropped from refresh-candidates (if that happens because something was already refreshed, the monitoring was already aborted so it's a no-op). 

I also changed the abort channel buffer to be larger than 1 because we can now have several writers to it. The state locks make it very unlikely but I think it's possible for a ``--ignore-running` refresh to cause a write to the abort channel, followed by `refreshhints.go` getting the state lock before the reader, writing to the channel again and deadlocking. I was thinking about possible tests for this but it seems quite difficult since it depends the timing of different handlers and goroutines getting the state lock.

File: overlord/snapstate/handlers.go [hunk 8]
```diff
@@ -948,14 +961,18 @@ func addMonitoring(st *state.State, snapName string, abort chan<- bool) error {
 
 	abortChans, ok := storedChans.(map[string]chan<- bool)
 	if !ok {
+		// reset the refresh hint since we're aborting
+		refreshHints[snapName].Monitored = false
+		st.Set("refresh-candidates", refreshHints)
+
 		// should never happen save for programmer error
-		return fmt.Errorf(`"monitored-snaps" should be map[string]chan<- bool but got %T`, storedChans)
+		return false, fmt.Errorf(`"monitored-snaps" should be map[string]chan<- bool but got %T`, storedChans)
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:972]: this should be marked internal error I suppose

File: overlord/snapstate/handlers.go [hunk 9]
```diff
@@ -915,7 +915,10 @@ func asyncRefreshOnSnapClose(st *state.State, snapsup *SnapSetup, refreshInfo *u
 		return fmt.Errorf("cannot monitor for snap closure: %w", err)
 	}
 
-	abort := make(chan bool, 1)
+	// use a buffer larger than 1 because there can be more than one writes to the channel.
+	// State locks make it unlikely but it's possible to have a second write before
+	// the reader can drain and remove the channel from the state
+	abort := make(chan bool, 16)
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:921]: we should probably switch to a cancellable context instead

File: overlord/snapstate/refreshhints.go [hunk 10]
```diff
@@ -97,10 +97,35 @@ func (r *refreshHints) refresh() error {
 	if err != nil {
 		return fmt.Errorf("internal error: cannot get refresh-candidates: %v", err)
 	}
+
+	stopMonitoringOutdatedCandidates(r.state, hints)
+
 	r.state.Set("refresh-candidates", hints)
```
Comment (pedronis) [inline overlord/snapstate/refreshhints.go:103]: do we need this logic in other places that set refresh-candidates ?

**Review by pedronis — COMMENTED** [review]
thanks for the latest set of changes, some comments, I'll try to take care of them

File: overlord/snapstate/refreshhints.go (hunk 10 above)
Comment (pedronis) [reply overlord/snapstate/refreshhints.go:103]: created a helper for this to use consistently

File: overlord/snapstate/handlers.go (hunk 8 above)
Comment (pedronis) [reply overlord/snapstate/handlers.go:972]: I shared more code around this and addressed this that way

**Review by ernestl — APPROVED** [review]
LGTM
